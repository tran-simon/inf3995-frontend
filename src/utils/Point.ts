export default interface Point {
  x: number;
  y: number;
}

export const newPoint = (x = 0, y = 0) => {
  return { x, y };
};

export const addPoint = (point1: Point, point2: Point | number): Point => {
  if (typeof point2 === 'number') {
    return addPoint(point1, newPoint(point2, point2));
  }
  return newPoint(point1.x + point2.x, point1.y + point2.y);
};

export const scalePoint = (point1: Point, point2: Point | number): Point => {
  if (typeof point2 === 'number') {
    return scalePoint(point1, newPoint(point2, point2));
  }
  return newPoint(point1.x * point2.x, point1.y * point2.y);
};
