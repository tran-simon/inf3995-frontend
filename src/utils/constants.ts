export const DEFAULT_ZOOM = 100;
