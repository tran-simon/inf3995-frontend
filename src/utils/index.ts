export * as Point from './Point';
export * as constants from './constants';
export * from './utils';
