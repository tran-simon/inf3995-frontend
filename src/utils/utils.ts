import { Dispatch, SetStateAction } from 'react';

export type SetState<T> = Dispatch<SetStateAction<T>>;

export type Point = {
  x: number;
  y: number;
};
