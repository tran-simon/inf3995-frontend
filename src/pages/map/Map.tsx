import {
  Dialog,
  DialogTitle,
  Grid,
  IconButton,
  Slider,
  Tooltip,
} from '@material-ui/core';
import React, { useContext, useRef, useState } from 'react';
import MapViewport from './MapViewport';
import CameraControls from './CameraControls';
import { useMouseHandler } from '../../hooks';
import { DEFAULT_ZOOM } from '../../utils/constants';
import Point, { addPoint, newPoint } from '../../utils/Point';
import SaveAltIcon from '@material-ui/icons/SaveAlt';
import CFContext from '../../context/CFContext';

export const calcZoom = (zoom: number, defaultZoom = DEFAULT_ZOOM) => {
  return (defaultZoom * defaultZoom) / zoom;
};

/**
 *
 * based on https://css-tricks.com/creating-a-panning-effect-for-svg/
 *
 */
const Map = () => {
  const mapRef = useRef<SVGSVGElement | null>(null);
  const [link, setLink] = useState<null | string>(null);
  const { name } = useContext(CFContext);

  /**
   * Zoom percentage
   */
  const [zoom, setZoom] = useState(DEFAULT_ZOOM);
  const size = calcZoom(zoom);

  const [cameraPos, setCameraPos] = useState<Point>(newPoint(0, 0));

  const ratioX =
    mapRef.current &&
    size / (mapRef.current.getBoundingClientRect().width || 1);
  const ratioY =
    mapRef.current &&
    size / (mapRef.current.getBoundingClientRect().height || 1);

  const mouseHandler = useMouseHandler({
    onMouseDrag: (event, origin) => {
      if (ratioX && ratioY) {
        const x = cameraPos.x - (event.clientX - origin.x) * ratioX;
        const y = cameraPos.y - (event.clientY - origin.y) * ratioY;
        setCameraPos(newPoint(x, y));
      }
    },
  });

  return (
    <>
      <MapViewport
        viewBox={`${cameraPos.x} ${cameraPos.y} ${size} ${size}`}
        style={{ position: 'absolute' }}
        ref={mapRef}
        onWheel={(event) => {
          if (event.ctrlKey) {
            setZoom(zoom - event.deltaY);
            event.preventDefault();
          } else {
            const deltaX = !ratioX
              ? 0
              : !event.shiftKey
              ? event.deltaX * ratioX
              : event.deltaY * ratioX;
            const deltaY =
              event.shiftKey || !ratioY ? 0 : event.deltaY * ratioY;

            setCameraPos(addPoint(cameraPos, newPoint(deltaX, deltaY)));
          }
        }}
        {...mouseHandler}
      />

      <Grid
        container
        style={{ height: '100%', padding: '1rem' }}
        alignContent="space-between"
      >
        <Grid item xs={11} container spacing={2}>
          <Grid item xs={12} sm={4}>
            <Slider
              id="slider-zoom"
              value={zoom}
              onChange={(e, v) => setZoom(v as number)}
              max={200}
              min={1}
            />
          </Grid>
        </Grid>
        <Grid item xs={1}>
          <IconButton
            style={{
              position: 'absolute',
              right: 0,
            }}
            onClick={() => {
              if (mapRef.current) {
                const serializer = new XMLSerializer();
                const source = serializer.serializeToString(mapRef.current);
                const url =
                  'data:image/svg+xml;charset=utf-8,' +
                  encodeURIComponent(source);
                setLink(url);
              }
            }}
          >
            <Tooltip title="Télécharger l'image SVG">
              <SaveAltIcon />
            </Tooltip>
          </IconButton>
        </Grid>
        <Grid
          item
          xs={12}
          sm={4}
          container
          spacing={2}
          style={{ height: '50%' }}
        >
          <CameraControls cameraPos={cameraPos} setCameraPos={setCameraPos} />
        </Grid>
      </Grid>
      <Dialog
        open={!!link}
        onClose={() => {
          setLink(null);
        }}
      >
        <DialogTitle>
          {link && (
            <a href={link} download={name + '.svg'}>
              Cliquez ici pour télécharger
            </a>
          )}
        </DialogTitle>
      </Dialog>
    </>
  );
};

export default Map;
