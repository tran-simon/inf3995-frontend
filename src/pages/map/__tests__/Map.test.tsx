import Map, { calcZoom } from '../Map';
import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { DEFAULT_ZOOM } from '../../../utils/constants';

describe('Map', () => {
  const delta = 10;

  it('can match snapshot', async () => {
    await waitFor(() => {
      expect(render(<Map />).asFragment()).toMatchSnapshot();
    });
  });

  it('can drag map using mouse', () => {
    render(<Map />);

    const viewport = document.querySelector('#map-viewport');
    if (viewport) {
      fireEvent.mouseDown(viewport, {
        clientX: 0,
        clientY: 0,
      });
      fireEvent.mouseMove(viewport, {
        clientX: delta,
        clientY: delta,
      });
      fireEvent.mouseUp(viewport);
    }

    const pos = -DEFAULT_ZOOM * delta;
    expect(viewport).toHaveAttribute(
      'viewBox',
      `${pos} ${pos} ${DEFAULT_ZOOM} ${DEFAULT_ZOOM}`,
    );
  });

  it('can zoom using mouse wheel', () => {
    render(<Map />);

    const viewport = document.querySelector('#map-viewport');

    if (viewport) {
      fireEvent.wheel(viewport, {
        ctrlKey: true,
        deltaY: 20,
      });
    }

    const zoom = calcZoom(80);
    expect(viewport).toHaveAttribute('viewBox', `0 0 ${zoom} ${zoom}`);
  });
});
