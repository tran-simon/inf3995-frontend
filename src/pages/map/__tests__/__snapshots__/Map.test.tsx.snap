// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map can match snapshot 1`] = `
<DocumentFragment>
  <svg
    height="100%"
    id="map-viewport"
    style="background-color: grey; position: absolute;"
    viewBox="0 0 100 100"
    width="100%"
  >
    <defs>
      <filter
        filterUnits="userSpaceOnUse"
        id="f1"
        x="0"
        y="0"
      >
        <fegaussianblur
          in="SourceGraphic"
          stdDeviation="5"
        />
      </filter>
    </defs>
  </svg>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-align-content-xs-space-between"
    style="height: 100%; padding: 1rem;"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item MuiGrid-grid-xs-11"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4"
      >
        <span
          class="MuiSlider-root MuiSlider-colorPrimary"
          id="slider-zoom"
        >
          <span
            class="MuiSlider-rail"
          />
          <span
            class="MuiSlider-track"
            style="left: 0%; width: 49.74874371859296%;"
          />
          <input
            type="hidden"
            value="100"
          />
          <span
            aria-orientation="horizontal"
            aria-valuemax="200"
            aria-valuemin="1"
            aria-valuenow="100"
            class="MuiSlider-thumb MuiSlider-thumbColorPrimary"
            data-index="0"
            role="slider"
            style="left: 49.74874371859296%;"
            tabindex="0"
          />
        </span>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root"
        style="position: absolute; right: 0px;"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            title="Télécharger l'image SVG"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 12v7H5v-7H3v7c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2v-7h-2zm-6 .67l2.59-2.58L17 11.5l-5 5-5-5 1.41-1.41L11 12.67V3h2z"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4"
      style="height: 50%;"
    >
      <div
        class="MuiBox-root MuiBox-root-8 makeStyles-root-6"
      >
        <div
          class="MuiBox-root MuiBox-root-9 makeStyles-leftSection-7"
        >
          <span
            class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-vertical"
          >
            <span
              class="MuiSlider-rail"
            />
            <span
              class="MuiSlider-track"
              style="bottom: 0%; height: 50%;"
            />
            <input
              type="hidden"
              value="0"
            />
            <span
              aria-orientation="vertical"
              aria-valuemax="100"
              aria-valuemin="-100"
              aria-valuenow="0"
              class="MuiSlider-thumb MuiSlider-thumbColorPrimary"
              data-index="0"
              role="slider"
              style="bottom: 50%;"
              tabindex="0"
            />
          </span>
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 8c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm8.94 3c-.46-4.17-3.77-7.48-7.94-7.94V1h-2v2.06C6.83 3.52 3.52 6.83 3.06 11H1v2h2.06c.46 4.17 3.77 7.48 7.94 7.94V23h2v-2.06c4.17-.46 7.48-3.77 7.94-7.94H23v-2h-2.06zM12 19c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
                />
              </svg>
            </span>
          </button>
        </div>
        <span
          class="MuiSlider-root MuiSlider-colorPrimary"
        >
          <span
            class="MuiSlider-rail"
          />
          <span
            class="MuiSlider-track"
            style="left: 0%; width: 50%;"
          />
          <input
            type="hidden"
            value="0"
          />
          <span
            aria-orientation="horizontal"
            aria-valuemax="100"
            aria-valuemin="-100"
            aria-valuenow="0"
            class="MuiSlider-thumb MuiSlider-thumbColorPrimary"
            data-index="0"
            role="slider"
            style="left: 50%;"
            tabindex="0"
          />
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
